### PR TITLE
Feat: disable rope scaling for training, add yarn during export

### PR DESCRIPTION
**Pull request by h-guo18**
## What does this PR do?

**Type of change:** New feature, bug fix

**Overview:**

- **Disable rope scaling during EAGLE training** for better training stability. The default EAGLE config now uses `rope_type: "default"` (no scaling) during training instead of llama3-style rope scaling.
- **Add YaRN rope scaling during export** for improved long-context inference quality. When exporting a model trained with default rope, the exporter sets `rope_type: "yarn"` with `factor=32.0`.
- **Add `eagle_train_length` config field** to `EagleConfig`, which propagates `training_seq_len` as `original_max_position_embeddings` in the exported YaRN rope config.
- **Propagate `rope_theta`** from the eagle config's `rope_parameters` into the exported HF config (required for transformers 5.x compatibility where `rope_theta` lives inside `rope_scaling`).
- **Add `trust_remote_code` flag** to `main.py` and `launch_train.sh` (was previously hardcoded to `True`).
- **Fix single-node GPU detection** in `launch_train.sh` to use `torch.cuda.device_count()` (respects `CUDA_VISIBLE_DEVICES`) instead of `nvidia-smi` (counts physical GPUs). Multi-node path keeps `nvidia-smi`.
- **Conditional FSDP2**: only use `fsdp_config.json` (FSDP2) when transformers >= 5.0; fall back to FSDP1 otherwise.

## Usage

```python
# Training: rope scaling is disabled by default for stability
# Export: yarn rope scaling is automatically applied
mtsp.convert(model, [("eagle", {
    "eagle_train_length": training_args.training_seq_len,  # propagated to rope config at export
    ...
})])
```

## Testing

Validated on single-node and multi-node training runs.

## Before your PR is "*Ready for review*"

- **Is this change backward compatible?**: Yes — `eagle_train_length` defaults to 2048, `trust_remote_code` defaults to `True` in the script.
- **Did you write any new necessary tests?**: No
- **Did you add or update any necessary documentation?**: No
- **Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?**: No

### CODE REVIEW

File: modelopt/torch/export/plugins/hf_spec_export.py [hunk 1]
```diff
@@ -175,6 +175,15 @@ def _get_config_from_draft_or_base(key: str, model: nn.Module):
         if self.hf_quant_config is not None:
             template_config["quantization_config"] = self.hf_quant_config
 
+        # For long context quality, we disable rope scaling for training
+        # and set yarn during export for inference.
+        template_config["rope_scaling"] = {
+            "rope_type": "yarn",
+            "rope_theta": 10000,
+            "factor": 32.0,
```
Comment (benchislett) [inline modelopt/torch/export/plugins/hf_spec_export.py]: I'm not sure these are the best choices for rope theta and factor. I think these might depend on how long max_position_embeddings actually is.

Some testing may be required. Gpt Oss uses rope theta 150k, for example. This may be some tradeoff between short-context and long-context accuracy

File: modelopt/torch/export/plugins/hf_spec_export.py [hunk 2]
```diff
@@ -175,6 +175,15 @@ def _get_config_from_draft_or_base(key: str, model: nn.Module):
         if self.hf_quant_config is not None:
             template_config["quantization_config"] = self.hf_quant_config
 
+        # For long context quality, we disable rope scaling for training
+        # and set yarn during export for inference.
+        template_config["rope_scaling"] = {
+            "rope_type": "yarn",
+            "rope_theta": 10000,
```
Comment (benchislett) [inline modelopt/torch/export/plugins/hf_spec_export.py]: Pretty sure rope theta goes on the main config and not the rope scaling, and should be set the same for training/inference. Where did this template come from?

File: modelopt/torch/speculative/eagle/default_config.py [hunk 3]
```diff
@@ -19,14 +19,7 @@
...
     "position_embedding_type": "rope",
-    "rope_scaling": {
-        "factor": 8.0,
-        "low_freq_factor": 1.0,
-        "high_freq_factor": 4.0,
-        "original_max_position_embeddings": 8192,
-        "rope_type": "llama3",
-    },
-    "rope_theta": 500000.0,
+    "rope_scaling": {"rope_type": "default", "rope_theta": 10000},
```
Comment (benchislett) [inline modelopt/torch/speculative/eagle/default_config.py:24]: I think you can go further and actually just set rope scaling to null. Not sure if there's a difference in HF

Comment (h-guo18) [reply modelopt/torch/speculative/eagle/default_config.py:24]: Setting it to None triggers an error. We are using Llama definition from transformers 5.0 and it requires a rope type. "rope_type":"default" here will use the traditional rope without scaling. 

ref: https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L85

Comment (h-guo18) [reply modelopt/torch/speculative/eagle/default_config.py:24]: This is a transforemr verfsion difference. In transforemr 4.x it's ok to leave it None. We are using transformer 5 here.

File: modelopt/torch/export/plugins/hf_spec_export.py (hunk 2 above)
Comment (h-guo18) [reply modelopt/torch/export/plugins/hf_spec_export.py]: This is also a version difference. In transformers 4.x it stays outside, while in transformers 5 it has to be in "rope_scaling" field.

Ref:
transforemr 4.55: https://github.com/huggingface/transformers/blob/v4.55-release/src/transformers/modeling_rope_utils.py#L110

5.x: 
https://github.com/huggingface/transformers/blob/main/src/transformers/modeling_rope_utils.py#L634
https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L111

File: modelopt/torch/export/plugins/hf_spec_export.py (hunk 1 above)
Comment (h-guo18) [reply modelopt/torch/export/plugins/hf_spec_export.py]: theta=10k is the default from HF: [ref](https://github.com/huggingface/transformers/blob/main/src/transformers/modeling_rope_utils.py#L630)

Actually my guess is that it should match the theta used in training.

factor should be a tradeoff I think.

File: examples/speculative_decoding/launch_train.sh [hunk 4]
```diff
@@ -185,15 +197,17 @@ else
...
 fi
 
+FSDP_ARGS=""
 if [[ "$TOTAL_GPU" -gt 1 ]]; then
-  #Use FSDP2 when multi GPU available
-  FSDP_ARGS="--fsdp 'full_shard' --fsdp_config ${SCRIPT_DIR}/fsdp_config.json"
-else
-  #Otherwise, single GPU training
-  FSDP_ARGS=""
+  # Use FSDP when multi GPU available, default to FSDP1
```
Comment (benchislett) [inline examples/speculative_decoding/launch_train.sh:206]: I believe we discussed using DDP by default for single-node multi-gpu

File: modelopt/torch/export/plugins/hf_spec_export.py [hunk 5]
```diff
@@ -183,6 +183,21 @@ def _get_config_from_draft_or_base(key: str, model: nn.Module):
                     new_value = str(new_value).replace("torch.", "")
                 template_config[key] = new_value
 
+        # For long context quality, we disable rope scaling for training
+        # and set yarn during export for inference.
+        if self.model.eagle_config.rope_parameters["rope_type"] == "default":
+            template_config["rope_scaling"] = {
+                "rope_type": "yarn",
+                "factor": 32.0,
+                "original_max_position_embeddings": getattr(self.model, "eagle_train_length", 4096),
```
Comment (benchislett) [inline modelopt/torch/export/plugins/hf_spec_export.py]: I think raising an error makes more sense than using a default for when "eagle_train_length" is missing. Incorrect rope should be flagged early and fail loudly IMO

File: modelopt/torch/speculative/eagle/default_config.py
Comment (benchislett) [inline modelopt/torch/speculative/eagle/default_config.py:46]: Unrelated, but these should be moved into json/yaml files and managed separately if we are going to have default configs for each model recipe (which, IMO, we should)

Comment (benchislett) [inline modelopt/torch/speculative/eagle/default_config.py:83]: I assume we'll want to do the same thing to the Kimi K2 config here

**Review by benchislett — CHANGES_REQUESTED** [review]

File: modelopt/torch/speculative/eagle/default_config.py
Comment (h-guo18) [reply modelopt/torch/speculative/eagle/default_config.py:83]: This model definition comes from a older transformers version and therefore have different condfig naming. Will deprecate this kimi_k2 decoder type soon since there's no longer one-model constraint for same kv cache format in TRTLLM and SGLang.

File: modelopt/torch/export/plugins/hf_spec_export.py [hunk 6]
```diff
@@ -183,6 +183,24 @@ def _get_config_from_draft_or_base(key: str, model: nn.Module):
...
+        # For long context quality, we disable rope scaling for training
+        # and set yarn during export for inference.
+        eagle_train_length = getattr(self.model, "eagle_train_length", None)
+        if eagle_train_length is None:
+            raise ValueError("eagle_train_length is needed for rope scaling but not set.")
+        if self.model.eagle_config.rope_parameters["rope_type"] == "default":
+            template_config["rope_scaling"] = {
+                "rope_type": "yarn",
+                "factor": 32.0,
+                "original_max_position_embeddings": getattr(self.model, "eagle_train_length", 4096),
```
Comment (ChenhanYu) [inline modelopt/torch/export/plugins/hf_spec_export.py]: `eagle_train_length` is fetched with `getattr(..., None)` and raises if `None`, but on line 195 it's fetched again with `getattr(..., 4096)` as a fallback default. These are inconsistent — if you intend to require the field, the second `getattr` with a fallback is dead code and misleading. Should use the already-validated `eagle_train_length` variable instead:
```python
"original_max_position_embeddings": eagle_train_length,
```

File: modelopt/torch/export/plugins/hf_spec_export.py [hunk 7]
```diff
@@ -183,6 +183,24 @@ def _get_config_from_draft_or_base(key: str, model: nn.Module):
...
+            }
+
+        # In transformer 5.x, rope_theta is under rope_parameters, rather than main config
+        if not template_config.get("rope_theta"):
```
Comment (ChenhanYu) [inline modelopt/torch/export/plugins/hf_spec_export.py]: `template_config.get("rope_theta")` is falsy for `0` and `0.0`, not just `None`. Consider `if template_config.get("rope_theta") is None:` for precision.

File: modelopt/torch/export/plugins/hf_spec_export.py [hunk 8]
```diff
@@ -183,6 +183,24 @@ def _get_config_from_draft_or_base(key: str, model: nn.Module):
...
                 template_config[key] = new_value
 
+        # For long context quality, we disable rope scaling for training
+        # and set yarn during export for inference.
+        eagle_train_length = getattr(self.model, "eagle_train_length", None)
+        if eagle_train_length is None:
+            raise ValueError("eagle_train_length is needed for rope scaling but not set.")
+        if self.model.eagle_config.rope_parameters["rope_type"] == "default":
+            template_config["rope_scaling"] = {
+                "rope_type": "yarn",
```
Comment (ChenhanYu) [inline modelopt/torch/export/plugins/hf_spec_export.py]: The hardcoded `"factor": 32.0` for YaRN scaling has no configurability. As @benchislett noted, this may not be universally correct. Consider making it a config field or at least documenting why 32.0 was chosen.

File: modelopt/torch/speculative/eagle/default_config.py [hunk 9]
```diff
@@ -19,14 +19,8 @@
     "hidden_act": "silu",
     "torch_dtype": "bfloat16",
     "position_embedding_type": "rope",
-    "rope_scaling": {
```
Comment (ChenhanYu) [inline modelopt/torch/speculative/eagle/default_config.py:22]: `rope_theta` appears both inside the `rope_scaling` dict and as a top-level key. Is the duplication intentional for compatibility across transformers versions? If so, a comment explaining why would help.

File: examples/speculative_decoding/main.py [hunk 10]
```diff
@@ -60,6 +60,7 @@
 @dataclass
 class ModelArguments:
     model_name_or_path: str | None = field(default="TinyLlama/TinyLlama-1.1B-Chat-v1.0")
```
Comment (ChenhanYu) [inline examples/speculative_decoding/main.py:62]: `trust_remote_code` defaults to `False` here but the launch script defaults it to `True`. This mismatch means running `main.py` directly (without the script) changes behavior from the previous hardcoded `True`. Consider aligning the defaults or documenting the discrepancy.

File: examples/speculative_decoding/launch_train.sh [hunk 11]
```diff
@@ -126,13 +134,21 @@ set -x
 
 SCRIPT_DIR="$(dirname "$(readlink -f "$0")")"
 NUM_NODES=${NUM_NODES:-1}
-GPU_PER_NODE=${GPU_PER_NODE:-$(nvidia-smi --query-gpu=name --format=csv,noheader | wc -l)}
-TOTAL_GPU=$((NUM_NODES * GPU_PER_NODE))
-echo "Total GPUs: $TOTAL_GPU (NUM_NODES: $NUM_NODES, GPU_PER_NODE: $GPU_PER_NODE)"
+if [[ "$NUM_NODES" != 1 ]]; then
+  #Multi Node Training
```
Comment (ChenhanYu) [inline examples/speculative_decoding/launch_train.sh:138]: FSDP is now opt-in (`--fsdp=True`), defaulting to `False`. Previously, multi-GPU automatically used FSDP. This is a **behavioral change** — multi-GPU training without FSDP may OOM or behave differently. The PR description doesn't call this out. Worth noting in the description or changelog.

File: examples/speculative_decoding/launch_train.sh [hunk 12]
```diff
@@ -154,7 +170,7 @@ DP_SHARD_SIZE=${DP_SHARD_SIZE:-$((TOTAL_GPU/CP_SIZE))}
 LOG_STEPS=${LOG_STEPS:-100}
 DRAFT_VOCAB_CACHE=${DRAFT_VOCAB_CACHE:-""}
 MIX_HIDDEN_STATES=${MIX_HIDDEN_STATES:-"False"}
-
+FSDP=${FSDP:-"False"}
```
Comment (ChenhanYu) [inline examples/speculative_decoding/launch_train.sh:173]: Nit: missing trailing newline at end of file.

**Review by ChenhanYu — CHANGES_REQUESTED** [review]
This PR makes several changes to the EAGLE speculative decoding pipeline: disables rope scaling during training (switches to "default"), applies YaRN at export, adds `eagle_train_length` config, makes `trust_remote_code` configurable, fixes single-node GPU detection, and adds conditional FSDP2. The changes look reasonable but there are some issues with hardcoded values and inconsistent defaults (see inline comments). **This PR also needs tests** — at minimum for the new export rope scaling logic (`hf_spec_export.py`) and the `eagle_train_length` propagation path, since incorrect rope configs would silently degrade inference quality.

File: examples/speculative_decoding/main.py (hunk 10 above)
Comment (h-guo18) [reply examples/speculative_decoding/main.py:62]: addressed in https://github.com/NVIDIA/Model-Optimizer/pull/1052/changes

File: modelopt/torch/speculative/eagle/default_config.py
Comment (h-guo18) [reply modelopt/torch/speculative/eagle/default_config.py:46]: This config is for the draft model. I think we will depreacate this kimi k2 drafter soon.

**Comment by h-guo18** [comment]
> This PR makes several changes to the EAGLE speculative decoding pipeline: disables rope scaling during training (switches to "default"), applies YaRN at export, adds `eagle_train_length` config, makes `trust_remote_code` configurable, fixes single-node GPU detection, and adds conditional FSDP2. The changes look reasonable but there are some issues with hardcoded values and inconsistent defaults (see inline comments). **This PR also needs tests** — at minimum for the new export rope scaling logic (`hf_spec_export.py`) and the `eagle_train_length` propagation path, since incorrect rope configs would silently degrade inference quality.

New unit tests added

**Review by benchislett — CHANGES_REQUESTED** [review]
"factor" should be equal to the ratio of the target model's max_position_embeddings and the training max_position_embeddings. 32 is not a reasonable default in all cases, and may cause a mismatch.

Also: I can't find where you are actually setting `max_position_embeddings` in the training config. This will need to be set to the training max seq len, and then reset to the target model's max seq len during export.

**Comment by benchislett** [comment]
It would also be helpful to see some actual validation of this scaling in effect, such as the work I did for the [SPEED-Bench paper](https://d1qx31qr3h6wln.cloudfront.net/publications/SPEED_Bench_Paper.pdf) (see figure on final page).

**Comment by h-guo18** [comment]
moved to PR https://github.com/NVIDIA/Model-Optimizer/pull/1238
